### ISSUE TITLE
Metadata performance regression

### ISSUE THREAD

**Issue by steveblackburn** [mmtk/mmtk-core#261]
The new metadata interface is abstract and general.    However, its generality leads to a measurable performance regression, which can be seen in the code snippets and graph below.

The code snippets show the before and after for the fast path of the OpenJDK C2 implementation of a simple object remembering barrier (a single bit per object, which is checked at each pointer write).    The additional instructions exist to allow a lot of generality in or metadata implementation.

We should consider a re-design that allows GCs which do not need that generality not to have to pay the price of such generality (ie so that we can revert to the first implementation of the barrier in this case).

### Before
```c++
// tmp2 = * (char*) ((obj >> 6) + MMTK_HEAP_END)
__ movptr(tmp2, obj);
__ shrptr(tmp2, 6);
__ movptr(tmp3, (intptr_t) MMTK_HEAP_END);
__ movl(tmp2, Address(tmp3, tmp2));
// tmp3 = (obj >> 3) & 7
// tmp2 = tmp2 >> tmp3
__ movptr(tmp3, obj);
__ shrptr(tmp3, 3);
__ andptr(tmp3, 7);
__ movptr(tmp4, rcx);
__ movl(rcx, tmp3);
__ shrptr(tmp2);
__ movptr(rcx, tmp4);
// if ((tmp2 & 1) == 0) goto slowpath;
__ andptr(tmp2, 1);
__ cmpptr(tmp2, (int32_t) NULL_WORD);
__ jcc(Assembler::equal, done);
__ movptr(c_rarg0, obj);
__ call_VM_leaf_base(CAST_FROM_FN_PTR(address, MMTkBarrierRuntime::record_modified_node), 1);
__ bind(done);
```

### After
```c++
// tmp3 = SIDE_METADATA_BASE_ADDRESS + ((addr & ~CHUNK_MASK) >> SIDE_METADATA_WORST_CASE_RATIO_LOG);
__ movptr(tmp3, obj);
__ movptr(tmp2, ~CHUNK_MASK);
__ andptr(tmp3, tmp2);
__ shrptr(tmp3, SIDE_METADATA_WORST_CASE_RATIO_LOG);
__ movptr(tmp2, SIDE_METADATA_BASE_ADDRESS);
__ addptr(tmp3, tmp2);
// tmp2 = * (uint8_t*) (((obj & CHUNK_MASK) >> 6) + tmp3)
__ movptr(tmp2, obj);
__ andptr(tmp2, CHUNK_MASK);
__ shrptr(tmp2, 6);
__ movb(tmp2, Address(tmp3, tmp2));
// tmp3 = (obj >> 3) & 7
__ movptr(tmp4, tmp3);
__ movptr(tmp3, obj);
__ shrptr(tmp3, 3);
__ andptr(tmp3, 7);
// tmp2 = tmp2 >> tmp3
__ movptr(tmp4, rcx);
__ movl(rcx, tmp3);
__ shrptr(tmp2);
__ movptr(rcx, tmp4);
// if ((tmp2 & 1) != 1) goto slowpath;
__ andptr(tmp2, 1);
__ cmpptr(tmp2, 1);
__ jcc(Assembler::notEqual, done);
__ movptr(c_rarg0, obj);
__ call_VM_leaf_base(CAST_FROM_FN_PTR(address, MMTkObjectBarrierSetRuntime::record_modified_node_slow), 1);
__ bind(done);
```

Note the code before `// tmp3 = (obj >> 3) & 7` in each case.

The regression is clear in the graph below (yellow to green):
![image](https://user-images.githubusercontent.com/5601406/109756906-e530e180-7c3c-11eb-8427-d6ca997b7c46.png)

The graph below includes the no-barrier variant, for context.
![image](https://user-images.githubusercontent.com/5601406/109757137-4bb5ff80-7c3d-11eb-9525-b03f0268f3f4.png)

**Comment by javadamiri** [comment]
It seems a considerable amount of the additional instructions are related to having per-chunk metadata (`CHUNK_MASK`).
So, maybe it's worth if we try putting each of the global and policy-specific side metadata sequentially?
In that case the side metadata address calculation will be more similar to Wenyu's old implementation

**Comment by steveblackburn** [comment]
That's right.    The trick is to do that without loss of generality, and also being able to handle discontiguous spaces (at least for 32 bit).

**Comment by javadamiri** [comment]
In our previous discussions on white board, we talked about and alternative to the per-chunk design, where we directly use the object address only with shift operations and an addition operation to a constant base address.
We can extend that to having two constant base addresses for global and policy-specific side metadata.
The calculations will be similar to the ones we do for inside a chunk, and we will not need to do the additional maskings discussed above

**Comment by steveblackburn** [comment]
Yes, but this depends critically on how the spaces are organized in virtual memory.    If they're contiguous this can be made to work.    If they're discontiguous, I don't think we have much choice but to make policy-specific metadata chunk-based.   Nonetheless, we can always make global metadata free of the chunk structure.

**Comment by javadamiri** [comment]
Right, and the metadata used in the above case is global, so making global metadata free of the chunk structure may help with the performance regression

**Comment by steveblackburn** [comment]
Correct

**Comment by javadamiri** [comment]
For the policy-specific side metadata, we can assign a constant base address to each policy in the 64-bits.

For the 32-bits case, I can't think of a better solution than the chunk-based approach.
However, I think it's not a big problem because in a 32-bits system, a non-generic policy-specific side metadata implementation is also not free to use a big contiguous piece of memory for the whole address space (the performance regression we're currently seeing is compared to such a case).
In other words, the non-generic implementation will have the same restriction as we have

**Comment by steveblackburn** [comment]
@javadamiri and I discussed the following solution today.

1. The basic approach is to have N+1 metadata regions, each large enough to cover the whole data address space. (For example if we allow 32 TB of data address space and consider a worst case metadata ratio of 0.5, we end up with N+1 meta data regions covering 16 TB each).
   a. 1 of the metadata regions is for global metadata
   b. N of the metadata regions are for the N distinct policies supported by MMTk (note: *policy* not *space*).
2. For every type metadata (global and policy-specific), there is a direct mapping from data address to the respective metadata, with no 'chunking' of metadata.
3. For 32-bit machines, we adopt the above for global metadata, while using the existing (chunked) scheme for policy-specific metadata.


![IMG_8152](https://user-images.githubusercontent.com/5601406/109906609-eae9fe00-7cf4-11eb-9e97-c3e9721858b7.JPG)

**Comment by javadamiri** [comment]
For all global and policy-specific side metadata, we know the specification at compile time.
I haven't fully implemented it yet, but I think we don't need an address space length of `HEAP_SIZE * worst_case_ratio` for each policy.

This is because we know the actual maximum address space size for each policy-specific metadata in advance, and we always keep it for all policies (which require side metadata), regardless of the plan that is chosen at runtime.

**Comment by steveblackburn** [comment]
Agree.   Note that it needs to be the worst case from among the various global metadata specs (different plans will have different requirements).
